From 8e2b654b2a118c29b6d17f074bf52e1a8f8da014 Mon Sep 17 00:00:00 2001
From: "Hayataka@kitakar5525" <Hayataka@kitakar5525>
Date: Thu, 20 Jun 2019 16:19:15 +0900
Subject: [PATCH] ipts: add debugfs entries
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add three debugfs entries:
・ quiesce_io_cmd
・ clear_mem_window_cmd
・ debug
・ ipts_restart
---
 drivers/misc/ipts/ipts-dbgfs.c       | 148 +++++++++++++++++++++++++++
 drivers/misc/ipts/ipts-msg-handler.h |   1 +
 2 files changed, 149 insertions(+)

diff --git a/drivers/misc/ipts/ipts-dbgfs.c b/drivers/misc/ipts/ipts-dbgfs.c
index 1c5c92f7d..232ac9bd7 100644
--- a/drivers/misc/ipts/ipts-dbgfs.c
+++ b/drivers/misc/ipts/ipts-dbgfs.c
@@ -20,9 +20,13 @@
 #include "ipts-sensor-regs.h"
 #include "ipts-msg-handler.h"
 #include "ipts-state.h"
+#include "../mei/mei_dev.h"
 
 const char sensor_mode_fmt[] = "sensor mode : %01d\n";
 const char ipts_status_fmt[] = "sensor mode : %01d\nipts state : %01d\n";
+const char ipts_debug_fmt[] = ">> tdt : fw status : %s\n"
+							  ">> == DB s:%x, c:%x ==\n"
+							  ">> == WQ h:%u, t:%u ==\n";
 
 static ssize_t ipts_dbgfs_mode_read(struct file *fp, char __user *ubuf,
 						size_t cnt, loff_t *ppos)
@@ -112,6 +116,122 @@ static const struct file_operations ipts_status_dbgfs_fops = {
         .llseek = generic_file_llseek,
 };
 
+static ssize_t ipts_dbgfs_quiesce_io_cmd_write(struct file *fp, const char __user *ubuf,
+						size_t cnt, loff_t *ppos)
+{
+	ipts_info_t *ipts = fp->private_data;
+	bool result;
+	int rc;
+
+	rc = kstrtobool_from_user(ubuf, cnt, &result);
+	if (rc)
+		return rc;
+
+	if (!result)
+		return -EINVAL;
+
+	ipts_send_sensor_quiesce_io_cmd(ipts);
+
+	return cnt;
+}
+
+static const struct file_operations ipts_quiesce_io_cmd_dbgfs_fops = {
+        .open = simple_open,
+        .write = ipts_dbgfs_quiesce_io_cmd_write,
+        .llseek = generic_file_llseek,
+};
+
+static ssize_t ipts_dbgfs_clear_mem_window_cmd_write(struct file *fp, const char __user *ubuf,
+						size_t cnt, loff_t *ppos)
+{
+	ipts_info_t *ipts = fp->private_data;
+	bool result;
+	int rc;
+
+	rc = kstrtobool_from_user(ubuf, cnt, &result);
+	if (rc)
+		return rc;
+
+	if (!result)
+		return -EINVAL;
+
+	ipts_send_sensor_clear_mem_window_cmd(ipts);
+
+	return cnt;
+}
+
+static const struct file_operations ipts_clear_mem_window_cmd_dbgfs_fops = {
+        .open = simple_open,
+        .write = ipts_dbgfs_clear_mem_window_cmd_write,
+        .llseek = generic_file_llseek,
+};
+
+static ssize_t ipts_dbgfs_debug_read(struct file *fp, char __user *ubuf,
+						size_t cnt, loff_t *ppos)
+{
+	ipts_info_t *ipts = fp->private_data;
+	char dbg_info[1024];
+	int len = 0;
+
+	char fw_sts_str[MEI_FW_STATUS_STR_SZ];
+	u32 *db, *head, *tail;
+	intel_ipts_wq_info_t* wq_info;
+
+	wq_info = &ipts->resource.wq_info;
+
+	mei_fw_status_str(ipts->cldev->bus, fw_sts_str, MEI_FW_STATUS_STR_SZ);
+	// pr_info(">> tdt : fw status : %s\n", fw_sts_str);
+
+	db = (u32*)wq_info->db_addr;
+	head = (u32*)wq_info->wq_head_addr;
+	tail = (u32*)wq_info->wq_tail_addr;
+	// pr_info(">> == DB s:%x, c:%x ==\n", *db, *(db+1));
+	// pr_info(">> == WQ h:%u, t:%u ==\n", *head, *tail);
+
+	if (cnt < sizeof(ipts_debug_fmt) - 3)
+		return -EINVAL;
+
+	len = scnprintf(dbg_info, 1024, ipts_debug_fmt,
+							fw_sts_str,
+							*db, *(db+1),
+							*head, *tail);
+	if (len < 0)
+		return -EIO;
+
+	return simple_read_from_buffer(ubuf, cnt, ppos, dbg_info, len);
+}
+
+static const struct file_operations ipts_debug_dbgfs_fops = {
+        .open = simple_open,
+        .read = ipts_dbgfs_debug_read,
+        .llseek = generic_file_llseek,
+};
+
+static ssize_t ipts_dbgfs_ipts_restart_write(struct file *fp, const char __user *ubuf,
+						size_t cnt, loff_t *ppos)
+{
+	ipts_info_t *ipts = fp->private_data;
+	bool result;
+	int rc;
+
+	rc = kstrtobool_from_user(ubuf, cnt, &result);
+	if (rc)
+		return rc;
+
+	if (!result)
+		return -EINVAL;
+
+	ipts_restart(ipts);
+
+	return cnt;
+}
+
+static const struct file_operations ipts_ipts_restart_dbgfs_fops = {
+        .open = simple_open,
+        .write = ipts_dbgfs_ipts_restart_write,
+        .llseek = generic_file_llseek,
+};
+
 void ipts_dbgfs_deregister(ipts_info_t* ipts)
 {
 	if (!ipts->dbgfs_dir)
@@ -143,6 +263,34 @@ int ipts_dbgfs_register(ipts_info_t* ipts, const char *name)
                 goto err;
         }
 
+        f = debugfs_create_file("quiesce_io_cmd", S_IWUSR, dir,
+                                ipts, &ipts_quiesce_io_cmd_dbgfs_fops);
+        if (!f) {
+                ipts_err(ipts, "debugfs quiesce_io_cmd creation failed\n");
+                goto err;
+        }
+
+        f = debugfs_create_file("clear_mem_window_cmd", S_IWUSR, dir,
+                                ipts, &ipts_clear_mem_window_cmd_dbgfs_fops);
+        if (!f) {
+                ipts_err(ipts, "debugfs clear_mem_window_cmd creation failed\n");
+                goto err;
+        }
+
+        f = debugfs_create_file("debug", S_IRUSR, dir,
+                                ipts, &ipts_debug_dbgfs_fops);
+        if (!f) {
+                ipts_err(ipts, "debugfs debug creation failed\n");
+                goto err;
+        }
+
+		f = debugfs_create_file("ipts_restart", S_IWUSR, dir,
+                                ipts, &ipts_ipts_restart_dbgfs_fops);
+        if (!f) {
+                ipts_err(ipts, "debugfs ipts_restart creation failed\n");
+                goto err;
+        }
+
 	ipts->dbgfs_dir = dir;
 
 	return 0;
diff --git a/drivers/misc/ipts/ipts-msg-handler.h b/drivers/misc/ipts/ipts-msg-handler.h
index 15038814d..f37d9ad9a 100644
--- a/drivers/misc/ipts/ipts-msg-handler.h
+++ b/drivers/misc/ipts/ipts-msg-handler.h
@@ -28,5 +28,6 @@ int ipts_send_feedback(ipts_info_t *ipts, int buffer_idx, u32 transaction_id);
 int ipts_send_sensor_quiesce_io_cmd(ipts_info_t *ipts);
 int ipts_send_sensor_hid_ready_for_data_cmd(ipts_info_t *ipts);
 int ipts_send_sensor_clear_mem_window_cmd(ipts_info_t *ipts);
+int ipts_restart(ipts_info_t *ipts);
 
 #endif /* _IPTS_MSG_HANDLER_H */
-- 
2.22.0

